openvscode-server: init at 1.62.0

dguenther opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

The derivation is based on the work done for [code-server](https://github.com/NixOS/nixpkgs/blob/17f230461a1010cdb0d0bb92594546429c69de47/pkgs/servers/code-server/default.nix), but is structured as a fork of upstream VS Code instead of a series of nested NPM packages.

ccing @colemickens as the creator of #140344

Closes #140344

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

dguenther committed:
openvscode-server: init at 1.62.0

dguenther commented:
Putting this in draft for a bit, I bumped the version to 1.62 and need to fix some build steps.

ghuntley commented on pkgs/servers/openvscode-server/default.nix:
```diff
@@ -0,0 +1,187 @@
+    outputHashMode = "recursive";
+    outputHashAlgo = "sha256";
+
+    # to get hash values use nix-build -A vscode-webserver.prefetchYarnCache
```
```suggestion
    # to get hash values use nix-build -A openvscode-server.prefetchYarnCache
```

ghuntley commented on pkgs/servers/openvscode-server/default.nix:
```diff
@@ -0,0 +1,187 @@
+  meta = with lib; {
+    description = "Run VS Code on a remote server";
+    longDescription = ''
+      vscode-webserver is VS Code running on a remote server, accessible through the
```
```suggestion
      openvscode-server is VS Code running on a remote server, accessible through the
```

ghuntley commented on pkgs/servers/openvscode-server/default.nix:
```diff
@@ -0,0 +1,187 @@
+    '';
+    homepage = "https://github.com/gitpod-io/openvscode-server";
+    license = licenses.mit;
+    maintainers = with maintainers; [ offline ghuntley ];
```
✅ Happy to help with maintenance, thanks for fixing the solution to esbuild post install scripts.

dguenther replied:
👍 @offlinehacker I removed you as maintainer from this new package so as not to force the effort of maintenance on you -- happy to add you back if you'd like!

dguenther commented on pkgs/servers/openvscode-server/default.nix:
```diff
@@ -0,0 +1,183 @@
+    # skip unnecessary electron download
+    export ELECTRON_SKIP_BINARY_DOWNLOAD=1
+  '' + lib.optionalString stdenv.isLinux ''
+
```
I think this comment was to remove whitespace? Should be updated now.

dguenther commented on pkgs/servers/openvscode-server/default.nix:
```diff
@@ -0,0 +1,183 @@
+
+  passthru = {
+    prefetchYarnCache = lib.overrideDerivation yarnCache (d: {
+      outputHash = lib.fakeSha256;
```
I deleted this section, from what I can tell it's not a common pattern in other packages.

dguenther commented:
I simplified the build process a bit and removed some unused dependencies, in addition to fixing @parcel/watcher on darwin -- this should be ready for a re-review.

colemickens commented:
@dguenther I'm sitting through a massive re-build, anything I should know for testing this? I was using this with code-server: https://github.com/cole-mickens/nixcfg/blob/43157e176fbf5914f31f7343cc29b9c24d7e4688/modules/code-server.nix (found somewhere online)

dguenther commented:
I haven't yet investigated the config options and different folders it creates for settings/extensions, but this might be a good starting point for config options: https://github.com/gitpod-io/openvscode-server/blob/main/src/vs/server/serverEnvironmentService.ts

The main path I've followed for testing has been to launch the server directly, load the URL it outputs in a browser, try opening the terminal (tests that native code was built properly), then install the Nix extension and try opening the `nixpkgs` git repository as a folder to test file watching. So if you have any ideas for additional testing beyond that, that'd be great!

colemickens commented:
Interesting, it seems like the auth is arguably both better and worse for this build than code-server. Better in that it feels like it actually enforces it a bit, and worse in that it's harder to just throw behind an auth'd proxy.

colemickens commented:
Aha, seems like the instructions for usage is to run `server.sh`, which lives in here with other web-server-launcher-goodies: https://github.com/gitpod-io/openvscode-server/tree/main/resources/server

colemickens commented:
I'll probably merge this soon, might be interesting to see if there's a number of us using this, maybe open a Matrix room and try to consolidate energy around one fork or the other.

dguenther commented:
> Interesting, it seems like the auth is arguably both better and worse for this build than code-server. Better in that it feels like it actually enforces it a bit, and worse in that it's harder to just throw behind an auth'd proxy.

Is that regarding the connection token? There were some notes about that token being new to 1.62, so that may end up in code-server eventually as well in some form, if they decide to use the upstream server: https://github.com/gitpod-io/openvscode-server#a-note-about-security-tokens

> Aha, seems like the instructions for usage is to run `server.sh`, which lives in here with other web-server-launcher-goodies: https://github.com/gitpod-io/openvscode-server/tree/main/resources/server

Yep, this file specifically is what gets output and is supposed to be executed: https://github.com/gitpod-io/openvscode-server/blob/main/resources/server/bin/server.sh

I'll see if I can update the derivation to run this script instead, that should minimize the chance of breakage in the future.

dguenther commented:
All right, updated it to run the `server.sh` script instead 👍

colemickens commented:
While I'm using this successfully and it roughly LGTM, and given that @ghuntley signed off on an earlier review makes me feel comfortable merging this.
